File: Week_06/121/best_time_to_buy_and_sell_stock.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) < 2:
            return 0
        min_price = prices[0]
        res = 0
        for p in prices[1:]:
            if p > min_price:
                res = max(res, p - min_price)
            else:
                min_price = p
        return res

File: Week_06/121/best_time_to_buy_and_sell_stock.go
package main

func maxProfit(prices []int) int {
	if len(prices) < 2 {
		return 0
	}
	res := 0
	minPrice := prices[0]
	for i := 1; i < len(prices); i++ {
		if prices[i] > minPrice {
			res = max(prices[i]-minPrice, res)
		} else {
			minPrice = prices[i]
		}
	}
	return res
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
